CI: Add linter job to CI workflow

TheodosiouTh opened the pull request:
Closes #165

preslavmihaylov commented:
@TheodosiouTh it seems there are some more linting issues once I ran it locally.

Note that the github action you've just added doesn't run on this PR's CI so we have to make sure the check passes locally.

Pasting the output:
![image](https://user-images.githubusercontent.com/7642330/132119105-cf19e16d-6132-4b5e-9034-5a12ef799d6d.png)

TheodosiouTh commented:
@preslavmihaylov Thank you for your input.
The pr is still a "work in progress" as I could not get it done yesterday.

preslavmihaylov commented:
Also, could you tweak the linter config a bit and include the `golint` and `gofmt` linters which are disabled by default.

Additionally, once you make that change, you can remove the gofmt step from the CI pipeline as its a subset of the step you're adding.

preslavmihaylov commented:
Actually, we should use `revive` as it seems `golint` is deprecated

TheodosiouTh commented:
@preslavmihaylov  I added `revive` and  `gofmt`  as you requested. I also removed the `gofmt` step and fixed all the linting errors and warnings. 
[Here is the final run of the workflow](https://github.com/TheodosiouTh/todocheck/runs/3517172940). 
The tests fail but I don't know why (on this PR they pass). I will keep looking at it to see what the issue is.

TheodosiouTh commented:
@preslavmihaylov  I have checked the code again and I do not get why the `make test` would fail after all the tests pass. My guess is that the error occurs because of the environment variables. So I think this will not be an issue when we merge the PR.

TheodosiouTh commented:
@preslavmihaylov  I think the PR is ready for review, feel free to check it and give me feedback.

preslavmihaylov commented on main.go:
```diff
@@ -77,7 +77,10 @@ func main() {
-		printTodoErrs(todoErrs, *format)
+		err = printTodoErrs(todoErrs, *format)
+		if err != nil {
+			logger.Info("Error when printing todo errors" + err.Error())
```
This should be a panic as we should never get here

TheodosiouTh replied:
I changed the code to panic if this point is reached during execution.

preslavmihaylov commented on testing/scenariobuilder/builder.go:
```diff
@@ -335,7 +336,10 @@ func (s *TodocheckScenario) setupMockIssueTrackerServer() (teardownFunc, error)
-				w.Write(issuetracker.BuildResponseFor(s.issueTracker, issue, s.issues[issue]))
+				_, err := w.Write(issuetracker.BuildResponseFor(s.issueTracker, issue, s.issues[issue]))
+				if err != nil {
+					logger.Info("Error " + err.Error() + " occured, when writing issuetracker response of issue:" + issue)
```
This should be a panic as it should never happen

TheodosiouTh replied:
I changed the code to panic if this point is reached during execution.

preslavmihaylov commented on traverser/todoerrs/todoerrs.go:
```diff
@@ -32,7 +33,10 @@ func commentsCallback(chk *checker.Checker, customTodos []string, todoErrCallbac
-			todoErrCallback(todoErr)
+			err = todoErrCallback(todoErr)
+			if err != nil {
+				logger.Info("couldn't run todo error callback: " + err.Error())
```
We should instead return `fmt.Errorf("received error from todo err callback: %w", err)`

TheodosiouTh replied:
I updated the code according to your suggestion.

preslavmihaylov commented on traverser/comments/comments.go:
```diff
@@ -42,14 +43,19 @@ func (t *Traverser) TraversePath(path string) error {
 		next = 0
-		t.handleStateChange(filename, line, linecnt, prev, curr, next)
+		err := t.handleStateChange(filename, line, linecnt, prev, curr, next)
+		if err != nil {
```
These errors shouldn't be handled here as they are handled via the last `return t.callbackErr`.

Does the linter allow ignoring the error via `_ = t.handleStateChange(...)`?

Alternatively, you can change the signature of the function to not return an `error` and update it accordingly

TheodosiouTh replied:
The linter does allow `_ = t.handleStageChange(...)` so I updated the code.

TheodosiouTh commented:
@preslavmihaylov I updated the code according to your suggestions. Also, I found out a logic bug in `printTodoErrs`, it was always returning an error even if an accepted format was provided (caused missing early return statements). I added the early return statements and changed the `if- else if` to a `switch-case` as I find switch case easier to read. You can see the bug fix on this (commit)[https://github.com/preslavmihaylov/todocheck/pull/166/commits/f0aac5aa40cb6262828b55cea10666d4e23dba88].

Also feel free to check the rest of the changes.

preslavmihaylov reviewed (approved):
Thank you for your contribution!
